test/topology_experimental_raft/test_tombstone_gc.py

```diff
@@ -198,3 +198,40 @@ async def verify_tombstone_gc(tombstone_mark, timeout=30):
 
             # test #3: the tombstones are cleaned up after the node is started again
             await verify_tombstone_gc(tombstone_mark)
+
+
+@pytest.mark.asyncio
+async def test_group0_state_id_failure(manager: ManagerClient):
+    """
+    Issue #21117 regression test.
+
+    Make sure that the "endpoint_state_map does not contain endpoint" warning is not triggered when the state_id is updated.
+
+    Arrange:
+        - start 1 node
+    Act:
+        - restart the node
+    Assert:
+        - the "endpoint_state_map does not contain endpoint" didn't appear in the log
+    """
+
+    # Start the node
+
+    srv = await manager.server_add()
+
+    log = await manager.server_open_log(srv.server_id)
+    mark = await log.mark()
+
+    cql = manager.get_cql()
+    await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60)
+
+    # Restart the node
+
+    await manager.server_restart(srv.server_id)
+    await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60)
+
+    # Check the log
+
+    with pytest.raises(TimeoutError):
+        await log.wait_for(r"Fail to apply application_state.*endpoint_state_map does not contain endpoint .* application_states = {GROUP0_STATE_ID",
+                           mark, timeout=1)
```

Comment:
Isn't this basically equivalent to `log.grep`?

Reply:
yes, we actually don't need to wait as we just check the past log (we don't even need the `mark` I think)

Reply:
done